server/mvcc/backend/backend.go

```diff
@@ -221,10 +238,45 @@ func (b *backend) ConcurrentReadTx() ReadTx {
 	defer b.readTx.RUnlock()
 	// prevent boltdb read Tx from been rolled back until store read Tx is done. Needs to be called when holding readTx.RLock().
 	b.readTx.txWg.Add(1)
+
 	// TODO: might want to copy the read buffer lazily - create copy when A) end of a write transaction B) end of a batch interval.
+	var modifiedBufVersion uint64
+	var buf *txReadBuffer
+
+	// we are using a mutex to protect the critical region where only one concurrent read tx can set the cache
+	b.cachedReadTxBuf.mu.Lock()
+	modifiedBufVersion = b.readTx.buf.Version()
+	if b.cachedReadTxBuf.buf == nil {
+		// initialize empty cache
+		bufCopy := b.readTx.buf.unsafeCopy()
+		b.cachedReadTxBuf.buf = &bufCopy
+		b.cachedReadTxBuf.bufVersion = modifiedBufVersion
+		buf = &bufCopy
+	} else {
+		// check if new data has been committed since last txReadBuffer copying.
+		if modifiedBufVersion != b.cachedReadTxBuf.bufVersion {
+			// we need to release the lock to maximize parallelism.
+			b.cachedReadTxBuf.mu.Unlock()
+			bufCopy := b.readTx.buf.unsafeCopy()
+			b.cachedReadTxBuf.mu.Lock()
+			buf = &bufCopy
+		} else {
+			buf = b.cachedReadTxBuf.buf
+		}
+		// it is possible we have buffer version updated during a unsafeCopy()
+		// if someone else has modified it, we will give up.
```

Comment:
Is this safe? If `modifiedBufVersion != b.readTx.buf.Version()`, we just use the stale cache?

Reply:
In this case, we just hold one exclusive copy of the readTxBuffer since someone has modified the data while we were doing a copy.